File: brain-move-web/src/games/puzzle.ts
import { BaseGame } from './base-game'

const EMOJI_POOL = [
  "\u{1F34E}", "\u{1F34A}", "\u{1F34B}", "\u{1F347}", "\u{1F349}", "\u{1F353}",
  "\u{1F33C}", "\u{1F33B}", "\u{1F33A}", "\u{1F337}", "\u{1F331}", "\u{1F340}",
  "\u2B50", "\u{1F535}", "\u{1F7E2}", "\u{1F534}", "\u{1F7E1}", "\u{1F7E3}",
  "\u{1F7E0}", "\u{1F7E4}", "\u{1F31F}", "\u{1F308}", "\u{1F300}", "\u{1F31E}",
]

export class Puzzle extends BaseGame {
  phase: "instruction" | "playing" = "instruction"
  gridSize = 2
  targetGrid: string[][] = []
  currentGrid: string[][] = []
  selectedR = -1
  selectedC = -1
  correctCount = 0
  totalMoves = 0
  hintCount = 0
  maxHints = 3
  instructionTimer = 0

  constructor(difficulty: any, scoring: any) {
    super(difficulty, scoring)
    this.name = "puzzle"
    this.displayName = "Puzzle"
  }

  start(): void {
    super.start()
    const level = this.difficulty.getLevel()
    this.gridSize = Math.min(2 + level, 6)
    this.maxHints = Math.max(1, 4 - level)
    this.phase = "instruction"
    this.selectedR = -1
    this.selectedC = -1
    this.correctCount = 0
    this.totalMoves = 0
    this.hintCount = 0
    this.instructionTimer = Date.now()
    this.feedbackMessage = `Puzzle ${this.gridSize}x${this.gridSize}: scambia due tessere per ricreare l'immagine`
  }

  update(_poseData?: any, _dt?: number): void {
    if (!this.running) return
    if (this.phase === "instruction" && (Date.now() - this.instructionTimer) / 1000 > 2) {
      this.phase = "playing"
      this._generatePuzzle()
    }
  }

  private _generatePuzzle(): void {
    this.targetGrid = Array.from({ length: this.gridSize }, () =>
      Array.from({ length: this.gridSize }, () => EMOJI_POOL[Math.floor(Math.random() * EMOJI_POOL.length)])
    )
    const flat = this.targetGrid.flat().sort(() => Math.random() - 0.5)
    this.currentGrid = Array.from({ length: this.gridSize }, (_, r) =>
      flat.slice(r * this.gridSize, (r + 1) * this.gridSize)
    )
  }

  selectPiece(r: number, c: number): void {
    if (this.phase !== "playing") return
    if (this.selectedR === -1) {
      this.selectedR = r; this.selectedC = c
      this.feedbackMessage = `Selezionato: riga ${r + 1}, colonna ${c + 1}. Scegli dove spostarlo`
    } else {
      if (this.selectedR === r && this.selectedC === c) {
        this.selectedR = -1; this.feedbackMessage = "Selezione annullata"; return
      }
      this._swap(this.selectedR, this.selectedC, r, c)
      this.selectedR = -1
    }
  }

  private _swap(r1: number, c1: number, r2: number, c2: number): void {
    const tmp = this.currentGrid[r1][c1]
    this.currentGrid[r1][c1] = this.currentGrid[r2][c2]
    this.currentGrid[r2][c2] = tmp
    this.totalMoves++
    this._checkProgress()
  }

  private _checkProgress(): void {
    let correct = 0
    for (let r = 0; r < this.gridSize; r++)
      for (let c = 0; c < this.gridSize; c++)
        if (this.currentGrid[r][c] === this.targetGrid[r][c]) correct++
    this.correctCount = correct
    const total = this.gridSize * this.gridSize
    if (correct === total) { this.feedbackMessage = "Puzzle completato!"; this._finishGame() }
    else this.feedbackMessage = `${correct}/${total} tessere al posto giusto`
  }

  useHint(): [number, number] | null {
    if (this.phase !== "playing" || this.hintCount >= this.maxHints) return null
    for (let r = 0; r < this.gridSize; r++)
      for (let c = 0; c < this.gridSize; c++)
        if (this.currentGrid[r][c] !== this.targetGrid[r][c]) {
          this.hintCount++
          this.feedbackMessage = `Indizio: la tessera in (${r + 1},${c + 1}) è sbagliata`
          return [r, c]
        }
    return null
  }

  private _finishGame(): void {
    this.stop()
    const penalty = 1 - (this.hintCount / Math.max(this.maxHints, 1)) * 0.15
    const efficiency = Math.max(0.3, 1 - this.totalMoves / Math.max(this.gridSize * this.gridSize * 3, 1))
    const accuracy = 0.7 * efficiency + 0.3 * penalty
    this.result = { completed: true, accuracy, correct: this.correctCount, totalMoves: this.totalMoves, gridSize: this.gridSize, hintsUsed: this.hintCount }
    this.difficulty.adapt(accuracy)
    this.scoring.addScore(this.name, Math.round(accuracy * 30))
  }

  getProgress(): number { return this.correctCount / Math.max(this.gridSize * this.gridSize, 1) }
  isSelected(r: number, c: number): boolean { return this.selectedR === r && this.selectedC === c }
}

File: brain-move/games/puzzle.py
import random
import time
import math
from games.base_game import BaseGame


class Puzzle(BaseGame):
    def __init__(self, difficulty_engine, scoring_engine):
        super().__init__(difficulty_engine, scoring_engine)
        self.name = "puzzle"
        self.display_name = "Puzzle"
        self.phase = "instruction"
        self.grid_size = 2
        self.EMOJI_POOL = [
            "\U0001F34E", "\U0001F34A", "\U0001F34B", "\U0001F347",
            "\U0001F349", "\U0001F353", "\U0001F33C", "\U0001F33B",
            "\U0001F33A", "\U0001F337", "\U0001F331", "\U0001F340",
            "\u2B50", "\U0001F535", "\U0001F7E2", "\U0001F534",
            "\U0001F7E1", "\U0001F7E3", "\U0001F7E0", "\U0001F7E4",
            "\U0001F31F", "\U0001F308", "\U0001F300", "\U0001F31E",
        ]
        self.target_grid = []
        self.current_grid = []
        self.selected_r = -1
        self.selected_c = -1
        self.correct_count = 0
        self.total_moves = 0
        self.hint_count = 0
        self.max_hints = 3
        self.instruction_timer = 0

    def start(self):
        super().start()
        level = self.difficulty.get_level()
        self.grid_size = min(2 + level, 6)
        self.max_hints = max(1, 4 - level)
        self.phase = "instruction"
        self.selected_r = -1
        self.selected_c = -1
        self.correct_count = 0
        self.total_moves = 0
        self.hint_count = 0
        self.instruction_timer = time.time()
        self.feedback_message = f"Puzzle {self.grid_size}x{self.grid_size}: scambia due tessere per ricreare l'immagine"

    def _generate_puzzle(self):
        self.target_grid = [[random.choice(self.EMOJI_POOL) for _ in range(self.grid_size)] for _ in range(self.grid_size)]
        flat = [self.target_grid[r][c] for r in range(self.grid_size) for c in range(self.grid_size)]
        random.shuffle(flat)
        self.current_grid = [flat[i * self.grid_size:(i + 1) * self.grid_size] for i in range(self.grid_size)]

    def update(self, pose_data=None, dt=0):
        if not self.running:
            return
        if self.phase == "instruction":
            if time.time() - self.instruction_timer > 2.0:
                self.phase = "playing"
                self._generate_puzzle()

    def select_piece(self, r, c):
        if self.phase != "playing":
            return
        if self.selected_r == -1:
            self.selected_r, self.selected_c = r, c
            self.feedback_message = f"Selezionato: riga {r+1}, colonna {c+1}. Scegli dove spostarlo"
        else:
            if self.selected_r == r and self.selected_c == c:
                self.selected_r = -1
                self.feedback_message = "Selezione annullata"
                return
            self._swap(self.selected_r, self.selected_c, r, c)
            self.selected_r = -1
            self.selected_c = -1

    def _swap(self, r1, c1, r2, c2):
        self.current_grid[r1][c1], self.current_grid[r2][c2] = self.current_grid[r2][c2], self.current_grid[r1][c1]
        self.total_moves += 1
        self._check_progress()

    def _check_progress(self):
        correct = 0
        for r in range(self.grid_size):
            for c in range(self.grid_size):
                if self.current_grid[r][c] == self.target_grid[r][c]:
                    correct += 1
        self.correct_count = correct
        total = self.grid_size * self.grid_size
        if correct == total:
            self.feedback_message = "Puzzle completato!"
            self._finish_game()
        else:
            self.feedback_message = f"{correct}/{total} tessere al posto giusto"

    def use_hint(self):
        if self.phase != "playing" or self.hint_count >= self.max_hints:
            return None
        for r in range(self.grid_size):
            for c in range(self.grid_size):
                if self.current_grid[r][c] != self.target_grid[r][c]:
                    self.hint_count += 1
                    self.feedback_message = f"Indizio: la tessera in ({r+1},{c+1}) e' sbagliata"
                    return (r, c)
        return None

    def _finish_game(self):
        self.stop()
        penalty = 1 - (self.hint_count / max(self.max_hints, 1)) * 0.15
        efficiency = max(0.3, 1 - self.total_moves / max(self.grid_size * self.grid_size * 3, 1))
        accuracy = (0.7 * efficiency + 0.3 * penalty)
        self.result = {
            "completed": True,
            "accuracy": accuracy,
            "correct": self.correct_count,
            "total_moves": self.total_moves,
            "grid_size": self.grid_size,
            "hints_used": self.hint_count,
        }
        self.difficulty.adapt(accuracy)
        self.scoring.add_score(self.name, int(accuracy * 30))

    def get_progress(self):
        total = self.grid_size * self.grid_size
        return self.correct_count / max(total, 1)

    def is_selected(self, r, c):
        return self.selected_r == r and self.selected_c == c
